x/checkpointing/prepare/transactions.go

```diff
@@ -0,0 +1,105 @@
+package prepare
+
+import (
+	"errors"
+	"fmt"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+)
+
+// PrepareProposalTxs is used as an intermediary storage for transactions when creating
+// a proposal for `PrepareProposal`.
+type PrepareProposalTxs struct {
+	// Transactions.
+	CheckpointTx []byte
+	OtherTxs     [][]byte
+
+	// Bytes.
+	// In general, there's no need to check for int64 overflow given that it would require
+	// exabytes of memory to hit the max int64 value in bytes.
+	MaxBytes  uint64
+	UsedBytes uint64
+}
+
+// NewPrepareProposalTxs returns a new `PrepareProposalTxs` given the request.
+func NewPrepareProposalTxs(
+	req *abci.RequestPrepareProposal,
+) (PrepareProposalTxs, error) {
+	if req.MaxTxBytes <= 0 {
+		return PrepareProposalTxs{}, errors.New("MaxTxBytes must be positive")
+	}
+
+	return PrepareProposalTxs{
+		MaxBytes:  uint64(req.MaxTxBytes),
+		UsedBytes: 0,
+	}, nil
+}
+
+// SetCheckpointTx sets the tx used for checkpoint
+func (t *PrepareProposalTxs) SetCheckpointTx(tx []byte) error {
+	oldBytes := uint64(len(t.CheckpointTx))
+	newBytes := uint64(len(tx))
+	if err := t.UpdateUsedBytes(oldBytes, newBytes); err != nil {
+		return err
+	}
+	t.CheckpointTx = tx
+	return nil
+}
+
+// AddOtherTxs adds txs to the "other" tx category and return early
+// if all the remaining bytes are used up
+func (t *PrepareProposalTxs) AddOtherTxs(allTxs [][]byte) error {
+	t.OtherTxs = make([][]byte, 0, len(allTxs))
+	bytesToAdd := uint64(0)
+	for _, tx := range allTxs {
+		txSize := uint64(len(tx))
+		if t.UsedBytes+bytesToAdd+txSize > t.MaxBytes {
+			break
+		}
+
+		bytesToAdd += txSize
+		t.OtherTxs = append(t.OtherTxs, tx)
+	}
+
+	if err := t.UpdateUsedBytes(0, bytesToAdd); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateUsedBytes updates the used bytes field. This returns an error if the num used bytes
+// exceeds the max byte limit.
+func (t *PrepareProposalTxs) UpdateUsedBytes(
```

Comment:
it seems this should be private, as using this without corresponding transactions operations would leave `PrepareProposalTxs` struct in bad state